pkg/gatewayserver/io/downlink_token.go

```diff
@@ -34,25 +40,49 @@ type DownlinkTokens struct {
 	items [downlinkTokenItems]downlinkToken
 }
 
-// Next returns a new downlink token.
-func (t *DownlinkTokens) Next(correlationIDs []string, time time.Time) uint16 {
+// Next returns a new downlink token for a downlink message.
+func (t *DownlinkTokens) Next(msg *ttnpb.DownlinkMessage, time time.Time) uint16 {
 	key := uint16(atomic.AddUint32(&t.last, 1))
 	pos := key % downlinkTokenItems
+
 	t.items[pos] = downlinkToken{
-		key:            key,
-		correlationIDs: correlationIDs,
-		time:           time,
+		key:  key,
+		msg:  msg,
+		time: time,
 	}
 	return key
 }
 
 // Get returns the correlation IDs and time difference between the time given to `Next` and the given time by the token.
-// If the token could not be found, this method returns false for `ok`.
-func (t DownlinkTokens) Get(token uint16, time time.Time) (correlationIDs []string, delta time.Duration, ok bool) {
+// If the token could not be found, this method returns false.
+func (t DownlinkTokens) Get(token uint16, time time.Time) (*ttnpb.DownlinkMessage, time.Duration, bool) {
 	pos := token % downlinkTokenItems
 	item := t.items[pos]
 	if item.key != token {
 		return nil, 0, false
 	}
-	return item.correlationIDs, time.Sub(item.time), true
+	return item.msg, time.Sub(item.time), true
+}
+
+var parseTokenRegex = regexp.MustCompile(`^gs:down:token:(\d+)$`)
+
+// FormatCorrelationID formats a correlation ID for a downlink token.
+func (t DownlinkTokens) FormatCorrelationID(token uint16) string {
+	return fmt.Sprintf("gs:down:token:%d", token)
+}
+
+// ParseTokenFromCorrelationIDs parses the correlation ID
+func (t DownlinkTokens) ParseTokenFromCorrelationIDs(cids []string) (uint16, bool) {
+	for _, cid := range cids {
+		match := parseTokenRegex.FindString(cid)
+		if match == "" {
+			continue
+		}
+		token, err := strconv.ParseUint(strings.TrimPrefix(match, "gs:down:token:"), 10, 16)
+		if err != nil {
```

Comment:
nit; you now use the `gs:down:token:` magic const twice. It's nicer to do a group regex match so you get the number directly